### ISSUE TITLE
Content negotiation: Look at accept header in request

### ISSUE THREAD

**Issue by faassen** [morepath/morepath#70]
Could we look at the accept header in request and create a predicate that switches on that?

**Comment by faassen** [comment]
I'm reopening this ticket for @estebistec

Clues on how to implement:

**Comment by faassen** [comment]
Another question is whether we should make the html and json directive do negotation already. This would move this into core territory. While that might happen at some point, I think at the least it needs to start as an extension.

**Comment by estebistec** [comment]
Taking a look at this now.

**Comment by estebistec** [comment]
I start with a simple example just to see how things should work and I wanted to make some comments and observations before I proceed. I want to see if they point to limitations by predicates and/or morepath and also give you a chance to react and see what you think.

First, the example morepath app I used: 

```python
# -*- coding: utf-8 -*

import morepath
import reg

from morepath import generic

class App(morepath.App):
    pass

@App.path(path='')
class Root(object):
    pass

@App.predicate(generic.view, name='accept', default=None,
               index=reg.KeyIndex,
               after=morepath.LAST_VIEW_PREDICATE)
def accept_header_predicate(request):
    return request.headers.get('Accept')

@App.html(model=Root, accept='text/html')
def hello_world(self, request):
    return "Hello world!"

@App.json(model=Root, accept='application/json')
def hello_world_json(self, request):
    return {'message': "Hello world!"}
```

Next, some simple curl-testing of this app

```bash
➜  ~ curl http://127.0.0.1:5000 -H 'Accept: text/html'
Hello world!
➜  ~ curl http://127.0.0.1:5000 -H 'Accept: application/json'
{"message": "Hello world!"}
➜  ~ curl http://127.0.0.1:5000 -H 'Accept: application/something-else'
404 Not Found

The resource could not be found.

➜  ~ curl http://127.0.0.1:5000
404 Not Found

The resource could not be found.

```

The first two commands activate the expected views. The second two commands are expected unknown (and unspecified).

So this points out the first problematic behavior. A default value only helps when a value isn't given from the request. How should failure-to-match be handled, so that the default `404 NOT FOUND` response can be overrided. In this case `406 NOT ACCEPTIBLE` is more appropriate.

Secondly, predicates seem to be tuned to match exact values, however content-types are not fixed values. They can have parameters, such as `charset` or `profile`. Further they can be more or less specific, such as `application/hal+json` or just `application/json`. This is just account for an individual media type. An accept header may have multiple such values, perhaps specified with q-values stating client preference or precedence.

For variable content-type resolution, predicates should be specifiable as lambdas instead of constant values. E.g., one could have `accept=lambda accept_value: accept_value.startswith("application/json")` (very dumbed-down) on the specific view method.

I'm still thinking about how to handle q-values. This is trickier because a value can only say whether it matches any specific entry in the accept header. E.g., "You wanted html with q=0.2, mine is more specific but I can cover that". So it would seem a hook into discriminator resolution may be needed (or if there is one I missed it). This piece of the topic is getting a little complex, and I don't want to suggest surgery on decate, predicates, etc. unless more features really are needed.

Thoughts?

**Comment by estebistec** [comment]
Currently reading through reg and dectate to see if I can get the feel for how this could/should work.

**Comment by faassen** [comment]
Reading ``core.py`` in Morepath might be of help. You can return a 406 error in a ``predicate_fallback ``directive.

It's true that fully supporting the Accept header is probably beyond the limits of what the predicate system can do. One reason behind this is because Reg is engineered to support caching, so that limits what kind of dynamic stuff you can do in there.

One suggestion is to consider your specific use cases instead of the full capabilities of the Accept header, and then to create a specific predicate that just pulls out the information you want into a fixed value (for instance, "is this asking for JSON or HTML") and let it tackle your use case.

You may be able to get a bit further with a specific Reg index. The ``ClassIndex`` is an example where a predicate can find a match in the registry for a class, also going through its base classes. It's possible you can tackle the q value system with this.

Note that WebOb offers some support in that it helps you test accept values: http://docs.webob.org/en/latest/reference.html#accept-headers

``list(request.accept)`` may possibly be of use.

It's conceivable that we want to add a second, slower style of predicate to Reg that lets you do more fancy stuff at the cost of caching, but that needs a lot more discussion.

Thanks for looking into this!

**Comment by estebistec** [comment]
Rather than spamming this issue with large tracts of code as a experiment I've created a running gist to show progress (we'll see how long this lasts).

https://gist.github.com/estebistec/b6667ecaea0ff20c4356ca6d8c5b1f58

The predicate fallback was successful, but I realized that client defaults are actually `*/*` (not really *no* accept header), in which case a default must be chosen instead of returning `406`.

**Comment by estebistec** [comment]
I'm pondering if the above (with some cleaning to disregard charsets, etc.) is a good enough start to be an installable *basic* content-negotiation feature. While I still have many concerns, something is better than nothing and there are many who would not invoke many of my concerns in basic applications (e.g., 99.999% of APIs won't offer *distinct* implementations representing different flavors of JSON). The notion of *[profiles](https://www.mnot.net/blog/2012/04/17/profiles)* is starting to gain some prevalence, but I don't expect it will eat very far into that 99% very soon.

How much more would *you* like to see from a *basic* option?

**Comment by estebistec** [comment]
As for fuzzy matching, I could do something clever like have the predicate return values with __eq__ overrided to provide the fuzziness. The problem here, even without q-values, is that a more general view could inadvertently pre-empt a more specific one (e.g., a view for plain JSON is matched before one for HAL, when HAL was asked for).

Even in the simpler situations of single-type accepts with variable specificity, it seems as if robustness and predictable behavior can only be achieved if there is some cumulative knowledge about the provided media types.

So, I'm going to take a look at indexes as you suggest, (and of course I'll absolutely re-use `webob.accept`) and see how far I can get.

**Comment by faassen** [comment]
Thank you so much for doing this experimentation!

Disclosure: I'm certainly no Accept header expert here. Reading up on q values now.. I'm certainly starting to understand the problem better now. I'm still not clear on what fuzzy matching is, though?

I think in Morepath code you should be able to write accept="something", where something is a simple media-type that can be used in an accept header. Like in your example. Maybe once we do the indexes right we can actually support writing accept="text/*" in views too, but we'll worry about that later.

For indexes, take a look at ``predicate.py`` in Reg, where ``ClassIndex`` subclasses from ``KeyIndex`` and implements permutations, in this case base classes, going from closest match to most distant match. 

We want our own subclass of KeyIndex that generates permutations the accept header we receive from the request. The idea is that the new index generates permutations of an accept header in order of preference. To get this right is going to get pretty  interesting. q values, plus specific over generic, etc. The predicate implementation itself can return "list(request.accept)"; reading webob.acceptparse it looks like it sorts by q value though at first glance doesn't prefer specific (``text/html``) over more general (``text/*``) so that would be a job for the index.

A more fundamental problem involves expanding stuff like ``text/*`` to actual media types like ``text/html``. We could build something fancy on top of some kind of registration database of media types (maybe there's a Python library we can reuse). This would be relatively slow as it'd need to cycle through all the "text/*" mimetypes, but that is probably fast enough as Morepath use Reg caching that skips all of that the second time the same combination of predicates is seen.

Alternatively we'd let the user create a specific index that knows about the media types possible in *their* application. That would make the generation of the specific media types a lot faster. Since making people install their own predicates with their own indexes is a bit cumbersome, we can wrap all that into a new directive that installs the proper predicate and index directly, i.e. something like:

```
@App.accept_media_types()
def get_media_types():
   return ["text/html", "application/json"]
```

but don't worry about that custom directive yet, the next goal now is to make our manual predicate with the custom index work.

**Comment by estebistec** [comment]
> I'm still not clear on what fuzzy matching is..

It's just my top of the head name for matching, eg, accept of HAL as plain
JSON

> and implements permutations

This is what I think will be hard. You are left choosing some arbitrary
constraints to expand on, because they are technically infinite for a given
type.

Also whatever we do will need to be per model.

Custom index that tracks offered types is what I'm currently playing with.
I'm just being overly verbose about minor observations along the way.

**Comment by estebistec** [comment]
Hrm... the email this morning totally destroyed the quotations that I was replying to, trying again...

> I'm still not clear on what fuzzy matching is..

It's just my top of the head name for matching, eg, accept of HAL as plain
JSON

> and implements permutations

This is what I think will be hard. You are left choosing some arbitrary
constraints to expand on, because they are technically infinite for a given
type.

Also whatever we do will need to be per model.

Custom index that tracks offered types is what I'm currently playing with.
I'm just being overly verbose about minor observations along the way.

**Comment by faassen** [comment]
To restrict the permutations I think it is a good idea to let the app say what kind of media types it supports. We could even deduce this from the predicates used by directives but recording this needs Reg support.

**Comment by estebistec** [comment]
Right. I don't think much of this is doable without some tracking.
Hopefully I'll have the next try in a few days.

**Comment by faassen** [comment]
For now I'd say focus on building an index factory that takes a set of allowed media types. We can do permutations on that list. Later we can see whether we can do something clever with configuration to automatically extract it. If Reg needs some work to make implementing this index more clean and easy I'd be happy to discuss ideas.

**Comment by estebistec** [comment]
Gist is updated with a _very_ _very_ basic hacky version that passes the original tests. I'll switch it over to properly use permutations, and see about proper model association next. There's a lot of tweaking and loose-ends but it's an okay start.

**Comment by faassen** [comment]
What do you mean by "model assocation"? The model class is associated with the view in another predicate so this should keep working.

Overriding ``get`` on the index is an interesting approach, I hadn't considered this as an option and I have to think about what it means.

**Comment by estebistec** [comment]
Yes, there may be nothing to do with models (aside from switch to a class
index). I'm just interested in doing some specific testing on it next time
I'm at this.

The get approach was simply the straightforward "see it work" approach. I
get the sense that the permutations method should take part of the job here.

The most significant aspect for me is that predicate values are webob mime
accepts instead of strings (I need to "fix" the default case). It seems to
offer some flexibility for the index lookup.

**Comment by faassen** [comment]
The other predicates (including the model with the class index) should just continue to work, so this should focus on matching the accept header only.

Concerning permutations, yes, I think it should create all permutations given the request's accept header information, the problem is simplified because we can determine what mime types the application itself supports.

**Comment by estebistec** [comment]
I'm struggling a bit with what you expect "permutations" to be at that point though. With an accept header you can pick a match at any point. I can do that in the permutations method, but it's never going to be more than one.

**Comment by estebistec** [comment]
Going to try to wrap up the basic implementation of this today.  The think the MIMEAccept class can take care of a reasonable chunk of possible use-cases. `best_match` will simply receive the list of keys available for a model (representing the offered types for that resource).

```
>>> from webob.acceptparse import MIMEAccept
'text/html'
>>> MIMEAccept('text/*').best_match(['text/html', 'text/plain'])
'text/html'
>>> MIMEAccept('text/*').best_match(['text/html', 'text/*'])
'text/html'
>>> MIMEAccept('text/plain').best_match(['text/html', 'text/*'])
'text/*'
>>> MIMEAccept('text/html').best_match(['text/html', 'text/*'])
'text/html'
>>> MIMEAccept('text/html;q=1, text/plain;q=0.5').best_match(['text/html', 'text/*'])
'text/html'
```

**Comment by estebistec** [comment]
Updated with examples of types like "text/*", and added another model to see that it continues to work for the separate URLs. 

The consequence of using MIMEAccepts as outputs of the predicate (when run against the request), is that the predicate values are not the same coming out of the predicate as those used to key possible views. This aspect does feel a bit hacky, and it does put the assumptions about how a predicate works under a slight mismatch. On the other hand I think the resulting predicate code is fairly clear and it does exactly what I expect content negotiation to do.

One final (fairly unrelated observation): the server seems to return HTTP/1.0 as the protocol, ignoring whatever the client stated would be used.

**Comment by taschini** [comment]
> One final (fairly unrelated) observation: the server seems to return HTTP/1.0 as the protocol, ignoring whatever the client stated would be used.

That's correct. This is the default behaviour of ``wsgiref.simple_server``, which is used by ``morepath.run``.  Keeping in mind that one should not use ``morepath.run`` in production, this has not been seen as an issue, up to now.  Feel free to open a separate issue if this behaviour is affecting your test setup or development environment.

**Comment by faassen** [comment]
Thanks for continuing with this. I am a bit confused you call ``self.d.keys()`` "offered_types", as they're not the types offered with the Accept header but they're instead the types *supported* by the program, correct? So isn't that an odd name? I think we need to add an API on indexes that will get you the list of all keys that were registered.

The reason I was thinking there would be more than one permutation is to implement the best matching, and have it fall back to a weaker match if the stronger one isn't supported by the program. What happens now if you have a new model and a view for 'text/*' while the request asks for 'text/html'? If there is any other explicit 'text/html' view in the system, it will try to find that view and it will fail to match the 'text/*' view. So I think this is a bug; I suspect it will turn up if you create a view for a new model 'Foo' and try to make it match with 'text/*'.

To make this more manageable I propose we turn this experiment into a proper ``more.accept`` project with some automated tests using pytest (with webtest). I've set up a completely empty repo in the Morepath project and added you as a collaborator; you should be able to push code. You can look at other ``more.`` libraries to see the structure of things; if you need help we're more than happy to help, though!

https://github.com/morepath/more.accept

**Comment by estebistec** [comment]
It's called "offers" because it is the list of alternatives offered by the
server. I don't recall where I picked up the terminology.

The model issue is something I'm still thinking about (it's what I
expressed concern for a bit earlier).

As for the specific example, I'm still thinking about it but it always
seemed odd to me that a resource representation wouldn't just say
explicitly what it was. But, other frameworks allow this and I understand
it's utility as a fallback so I'll add it to my test suite and work on it.

Thanks for creating the repo. I'll start populating it soon. I think that
not only will tests be important there but examples of how programmers
would write their own routing and mime type tests to verify that level of
their API. So I'm thinking a little about that too.

**Comment by faassen** [comment]
Ah, okay, I understand the 'offer' terminology better now, thank you.

I suspect due the model issue you cannot escape implementing 'permutations' to offer multiple possibilities (still based on what the server offers). This might mean we can reuse less of MimeAccept, though I haven't entirely given up hope there yet.

Glad you'll start filling the repo. I'm excited to see a ``more.accept`` library! I agree it needs a good README at the least that describes how you'd use it. I'm not sure it needs a lot on how a programmer would write tests -- the programmer would have to trust to a certain extent that ``more.accept`` does what it promises.

**Comment by estebistec** [comment]
It's not just trusting what accept does, it's ensuring their own routing
and set of offered resources/mime-types works after they, e.g., add a view
that offers "text/*". If you had a fully mocked API that did nothing with
input data, and gave dummy output data, you would still have API-specific
logic to verify.

**Comment by faassen** [comment]
Sure, hints on how to ensure this would be good. I think explaining how to do this with webtest would be great. Note that we now do have docs on how to test in general:

http://morepath.readthedocs.io/en/latest/testing.html
